### PR TITLE
HIVE-25534: Error when executing DistCp on file system not supporting XAttrs

### PR DESCRIPTION
**Pull request by hmangla98**

### CODE REVIEW

File: shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java [hunk 1]
```diff
@@ -1131,7 +1131,7 @@ public void setStoragePolicy(Path path, StoragePolicyValue policy)
       }
     }
     if (needToAddPreserveOption) {
-      params.add("-pbx");
+      params.add("-pb");  //Only Block Size will be preserved.
```
Comment (pkumarsinha) [inline shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java:1134]: We can't remove it in all the cases. for hdfs we should preserve

File: shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java [hunk 2]
```diff
@@ -1273,8 +1272,6 @@ public boolean runDistCpWithSnapshots(String oldSnapshot, String newSnapshot, Li
       }
     } catch (Exception e) {
       throw new IOException("Cannot execute DistCp process: ", e);
-    } finally {
-      conf.setBoolean("mapred.mapper.new-api", false);
```
Comment (pkumarsinha) [inline shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java:1277]: why is this change a part of this JIRA

Comment (pkumarsinha) [reply shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java:1277]: add test

File: shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java (hunk 1 above)
Comment (hmangla98) [reply shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java]: Change done.

File: shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java (hunk 2 above)
Comment (hmangla98) [reply shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java]: Added

File: shims/0.23/src/main/test/org/apache/hadoop/hive/shims/TestHadoop23Shims.java [hunk 3]
```diff
@@ -94,6 +97,26 @@ public void testConstructDistCpParams() {
 
   }
 
+  @Test
```
[the remark this answers is not in the document]
Comment (hmangla98) [reply shims/0.23/src/main/test/org/apache/hadoop/hive/shims/TestHadoop23Shims.java:100]: Test for non supported FS is already included in testConstructDistCpParams where we specify that XAttr are not preserved for LocalFileSystem.

[the remark this answers is not in the document]
Comment (pkumarsinha) [reply shims/0.23/src/main/test/org/apache/hadoop/hive/shims/TestHadoop23Shims.java:100]: I think you missed the word "other"  in above comment

File: shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java [hunk 4]
```diff
@@ -277,6 +277,18 @@ private boolean isFairScheduler (Configuration conf) {
...
   }
 
+  private boolean checkFileSystemXAttrSupport(FileSystem fs) throws IOException {
+    try {
+      fs.getXAttrs(new Path(Path.SEPARATOR));
+    } catch (UnsupportedOperationException e) {
+      LOG.warn("XAttr won't be preserved since it is not supported for file system: " + fs.getUri());
+      return false;
+    } catch (IOException e) {
+      throw e;
```
Comment (pkumarsinha) [inline shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java:287]: Why do you need this?

File: shims/0.23/src/main/test/org/apache/hadoop/hive/shims/TestHadoop23Shims.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (hmangla98) [reply shims/0.23/src/main/test/org/apache/hadoop/hive/shims/TestHadoop23Shims.java:102]: Included the test for S3AFileSystem.

**Review by pkumarsinha — COMMENTED** [review]
add check for src as well.

File: shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java [hunk 5]
```diff
@@ -1192,12 +1204,15 @@ public Boolean run() throws Exception {
...
-               .build();
+    DistCpOptions.Builder builder = new DistCpOptions.Builder(srcPaths, dst)
+            .withSyncFolder(true)
+            .withDeleteMissing(true)
+            .preserve(FileAttribute.BLOCKSIZE);
+    if (checkFileSystemXAttrSupport(dst.getFileSystem(conf))
+            && checkFileSystemXAttrSupport(srcPaths.get(0).getFileSystem(conf))) {
+      builder = builder.preserve(FileAttribute.XATTR);
+    }
+    DistCpOptions options = builder.build();
```
Comment (pkumarsinha) [inline shims/0.23/src/main/java/org/apache/hadoop/hive/shims/Hadoop23Shims.java:1215]: What is the difference between option and params returned by constructDistCpParams? Why are we passing preserve attribute in both. Can one not suffice?

**Review by pkumarsinha — APPROVED** [review]
LGTM, pending tests

**Comment by aasha** [comment]
Committing the same as the tests are not related.

**Pull request merged**